compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1938,10 +1948,48 @@ public void visit(BLangRecordDestructure recordDeStmt) {
             return;
         }
         typeChecker.checkExpr(recordDeStmt.expr, this.env);
+        if (checkValidMappingBindingPatternForRecordVarRef(recordDeStmt.varRef.recordRefFields,
+                recordDeStmt.expr.getBType())) {
+            dlog.error(recordDeStmt.expr.pos, DiagnosticErrorCode.INVALID_OPTIONAL_FIELD_IN_MAPPING_BINDING_PATTERN);
+        }
         checkRecordVarRefEquivalency(recordDeStmt.pos, recordDeStmt.varRef, recordDeStmt.expr.getBType(),
                                      recordDeStmt.expr.pos);
     }
 
+    private boolean checkValidMappingBindingPatternForRecordVarRef(List<BLangRecordVarRefKeyValue> recordRefFields,
+                                                                   BType type) {
+        type = types.getReferredType(type);
+        if (type.tag != TypeTags.RECORD) {
```

Comment:
We also need to handle maps, right?

https://ballerina.io/ballerina-spec/spec.html#mapping-binding-pattern

> a mapping-binding-pattern { f1: p1, f2: p2, ..., fn: pn, r } matches a mapping value m that includes fields f1, f2, ... , fn if pi matches the value of field fi for each i in 1 to n; if r is ...v then a successful match causes a new mapping value consisting of all the other fields to be assigned to v; a field-binding-pattern consisting of just a variable-name x is equivalent to a field-binding-pattern x: x;

So we can't really use a mapping binding-pattern in a destructuring assignment or variable declaration with a map as the expression?

Reply:
We currently allow samples like below right?
```bal
map<string> mapvar = {a: "my"};

public function main() {
    string? b;
    {a: b} = mapvar;
}
```
I think the spec doesn't specifically mention allowing or disallowing this but this can be useful right?

Reply:
Similar scenario, right? A map doesn't have required fields, so using a map with a mapping binding pattern that has field binding patterns (in a variable declaration or destructuring assignment) is not allowed by the spec?

Reply:
Yeah, right now this is allowed in this way,
```
    map<int> m = {
        "x": 1,
        "y": 2
    };

    var {x:x1, y:y1} = m;
```
Types of the `x1` and` y1` are` int?`. I think it is because map does not have required fields.

Reply:
Discussed to disallow mapping binding patterns that has an expression of map